docs/technologies.md

```diff
@@ -0,0 +1,423 @@
+                               internal_tables=INTERNAL_TABLES,
+                               external_tables=EXTERNAL_TABLES
+                               )\
+            .add_configuration('prod',
+                               project_id='my-project-prod')
+``` 
+  
+Having that, a `Dataset` object can be easily created:
+
+```python
+dataset = dataset_config.create_dataset()
+``` 
+
+Then, you can use short table names in SQL, a `Dataset` object resolves them to fully qualified names.
+
+For example, in this SQL, a short name of an internal table: 
+
+```python
+dataset.collect('select * from {quality_metric}')
+```
+
+is resolved to `my-project-dev.scorer.quality_metric`.
+
+In this SQL, an alias of an external table: 
+
+```python
+dataset.collect('select * from {offer_ctr}')
+```
+
+is resolved to `not-my-project.offer_scorer.offer_ctr_long_name`.
+
+### Dataset
+
+#### Definition
+
+A `Dataset` object has the following interface:
+
+**[`interface.py`](../bigflow/bigquery/interface.py)**
+
+```python
+class Dataset(object, metaclass=ABCMeta):
+
+    @abstractmethod
+    def write_truncate(self, table_name: str, sql: str, partitioned: bool = True) -> BigQueryOperation:
+        pass
+
+    @abstractmethod
+    def write_append(self, table_name: str, sql: str, partitioned: bool = True) -> BigQueryOperation:
+        pass
+
+    @abstractmethod
+    def write_tmp(self, table_name: str, sql: str) -> BigQueryOperation:
+        pass
+
+    @abstractmethod
+    def collect(self, sql: str) -> BigQueryOperation:
+        pass
+
+    @abstractmethod
+    def collect_list(self, sql: str) -> BigQueryOperation:
+        pass
+
+    @abstractmethod
+    def create_table(self, create_query: str) -> BigQueryOperation:
+        pass
+```
+
+All the methods are lazy and return a `BigQueryOperation` object, which is defined as the following interface:
+
+```python
+class BigQueryOperation(object, metaclass=ABCMeta):
+
+    @abstractmethod
+    def to_job(self, id: str, retry_count: int = DEFAULT_RETRY_COUNT, retry_pause_sec: int = DEFAULT_RETRY_PAUSE_SEC):
+        pass
+
+    def run(self, runtime=DEFAULT_RUNTIME):
+        pass
+```
+
+You can turn a lazy operation into a job or simply run it (useful for ad-hoc queries or debugging).
+
+A SQL code that you provide to the methods is templated (as mentioned in the previous section). 
+Besides a configuration parameters, you can access the
+`runtime` parameter. It's available as the `dt` variable. For example:
+
+```python
+dataset.write_truncate('target_table', '''
+SELECT *
+FROM `{another_table}`
+WHERE PARTITION_TIME = '{dt}'
+''')
+```
+
+#### Write truncate
+
+The `write_truncate` method takes a SQL query, executes it, and saves a result into a specified table.
+
+```python
+dataset.write_truncate('target_table', '''
+SELECT *
+FROM `{another_table}`
+''')
+```
+
+This method overrides the whole table or a single partition, depending on the type of a specified table.
```

Comment:
This method overrides the all data in a table or its single partition, depending on the `partitioned` parameter.

Reply:
done